### ISSUE TITLE
Only q4_k madlad400 model appears to be available

### ISSUE THREAD

**Issue by AppleSheeple** [CrispStrobe/CrispASR#333]
Hello.

https://huggingface.co/cstr/madlad400-3b-mt-GGUF/tree/main

Only the q4_k model appears to available, even though the README itself mentions f16 and q8_0.

**Comment by CrispStrobe** [comment]
Fixed — both files are up now:

- [`madlad400-3b-mt-f16.gguf`](https://huggingface.co/cstr/madlad400-3b-mt-GGUF/blob/main/madlad400-3b-mt-f16.gguf) (5.89 GB)
- [`madlad400-3b-mt-q8_0.gguf`](https://huggingface.co/cstr/madlad400-3b-mt-GGUF/blob/main/madlad400-3b-mt-q8_0.gguf) (3.38 GB)

You were right that the README was the problem, and it was worse than a stale table: the quickstart's own copy-paste lines told you to download `…-q8_0.gguf`, so following the card was guaranteed to 404. Only `-m auto` worked, because the registry pointed at `q4_k`.

Rather than just uploading the missing files, I used the occasion to give MADLAD a per-stage parity harness, because it turned out to have none — every published quant so far had been validated by "it loaded and the German looked right", which is not much of a check on a 3B encoder-decoder. There is now a reference dumper that walks the original `google/madlad400-3b-mt` checkpoint with HF's T5 semantics, and a `crispasr-diff madlad` arm that compares our runtime against it stage by stage.

**Per-stage cosine vs the PyTorch blueprint** — encoder embedding, position bias, four encoder blocks, encoder output, cross-attention K/V, decoder embedding, two decoder blocks, decoder output, and the step-0 logits:

| | worst cosine (14 stages) | step-0 argmax |
|---|---|---|
| **F16** | **1.000000** — every stage | MATCH |
| **Q8_0** | 0.999894 | MATCH |
| **Q4_K** | 0.993328 | MATCH |

F16 landing on 1.000000 across the board says the T5 port itself is faithful; the quants then degrade in the order you'd want, and all three still choose the same first token. `enc_pos_bias` is bit-exact (`max_abs=0.00000`) at every precision, which is the relative-position-bucket logic — bidirectional for the encoder, causal for the decoder — confirming itself; that is the piece of T5 that most often goes quietly wrong.

Each artifact was also validated by real translations before upload, not just a successful load:

```
en→de   "Hallo Welt, wie geht es dir heute?"
en→fr   "L'apprentissage automatique change le monde."
```

If you want to reproduce the table, the reference archive is in the same repo (`madlad400-3b-mt-ref.gguf`, 1.5 MB):

```bash
crispasr-diff madlad madlad400-3b-mt-q8_0.gguf madlad400-3b-mt-ref.gguf
```

**One thing worth flagging while you're using this**, since I found it on the way and it is not fixed: greedy decode does not always stop cleanly on very short inputs. One of my test cases, `"Bonjour le monde!"` fr→en, translated correctly to "Hello world!" and then ran on into a long repeat loop of digits. The translation is there, but you may want to cap `--max-new-tokens` on short inputs until that is dealt with properly. Tracked on our side.

Q4_K is unchanged and still fine — its numbers above are the first hard evidence of that, rather than an assurance. Thanks for the report; a missing file in a repo is easy to miss from the inside, and the quickstart pointing at it made it a real trap.

**Comment by CrispStrobe** [comment]
Following up on the decode caveat I flagged, because the answer turned out to be worth stating precisely.

I assumed the runaway decode was our bug — a port that reproduces a model's *outputs* can still get its *stopping* wrong, and that's the usual suspect. It isn't. I ran the PyTorch reference on the same input, and it runs away identically:

```
reference (HF T5, fp32):  60 tokens, ran to the cap, no EOS
                          'Hello world! – 100000000000000000000000000000…'
CrispASR:                 'Hello world! – 1000000000000000000000000'
```

while the two inputs it *does* terminate on match us exactly (`'Hallo Welt, wie geht es dir heute?'`, 10 tokens; `'L'apprentissage automatique change le monde.'`, 11 tokens). So this is MADLAD's own behaviour on some short inputs, not something the GGUF conversion or the ggml runtime introduced.

Ruling that out took one arm per candidate rather than a guess, and each is worth knowing if you hit something similar:

| suspect | ruled out by |
|---|---|
| quantization | reproduces at F16, where per-stage parity is 1.000000 |
| KV cache | re-forwarding the whole prefix every step gives the *identical* string |
| tokenizer | runtime independently emits the same 11 ids, ending in EOS=2 |
| EOS id | `config.json`, the GGUF and the runtime all say 2 |
| the graph | 14/14 stages pass, step-0 argmax matches |

**What changed on our side anyway:** greedy decode now detects a repeated token cycle and stops, trimming the degenerate tail — `"Hello world! – 10"` instead of `"…– 10000000000…"`. It's behind `CRISPASR_T5_REPEAT_BREAK=0` precisely *because* it deliberately deviates from the blueprint; if you're diffing our output against HF's, turn it off and you'll get the reference behaviour byte for byte. Both sentences that terminate cleanly are unchanged with it on or off.

So the earlier advice to cap `--translate-max-tokens` on short inputs is no longer needed, though it remains a reasonable belt-and-braces.
